### PR TITLE
Prepare for tested articles

### PR DESCRIPTION
**Pull request by TomasVotruba**
This PR adds support for tested articles.

Inspired and discused with @enumag 

All but one descision is done:

## Where to put test files and how to name them related to post? 

- See options 1, 2, 3 and 4 bellow:
- Support one or suggest new with reasoning.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,23 +2,52 @@ language: php
...
+
+# for section info see: https://docs.travis-ci.com/user/deployment
+# run everywhere excepts PRs
+deploy:
+  # we do it manually
+  provider: script
+  script:
     - vendor/bin/statie generate
-    - vendor/bin/statie push-to-github pehapkari/pehapkari.cz --token=${GH_TOKEN}
+    - vendor/bin/stati  e push-to-github pehapkari/pehapkari.cz --token=${GH_TOKEN}
```
Comment (enumag) [inline .travis.yml]: typo

**Comment by TomasVotruba** [comment]
LGTM means?

**Comment by enumag** [comment]
"Looks Good To Me" ;-)

**Comment by TomasVotruba** [comment]
Ah. That was too easy :)  Thanks

What about naming conventions. Do you have any ideas?

A. Name it 1:1:

- `source/_post/2016-12-05-symfony....md`
- `src/_posts/2016-12-05-symfony..../....php`
- `tests/_posts/2016-12-05-symfony..../....Test.php`

**Comment by enumag** [comment]
I've a slight problem with dates in filenames - as we discussed it might take a while to merge an article so they might not be accurate. On the other hand I assume they're needed for ordering, right?

**Comment by enumag** [comment]
Then there is a question what to do with consecutive articles that could share their code.

**Comment by enumag** [comment]
### 2. Drop the date from src and tests

    - `source/_posts/2016-12-05-symfony....md`
    - `src/_posts/symfony..../....php`
    - `tests/_posts/symfony..../....Test.php`


I'm not sure how often it will be needed to change a date of an article before merging. If common it would be better to have the date on less places.

**Comment by enumag** [comment]
Also I'd recommend to have a subdirectory per year as the number of articles could grow fast.

**Comment by TomasVotruba** [comment]
> Then there is a question what to do with consecutive articles that could share their code.

This we can figure out when that happens. Now it's about naming only.

> Also I'd recommend to have a subdirectory per year as the number of articles could grow fast.

Good point, I thought about it too.

**Comment by TomasVotruba** [comment]
For 1. and 2. location is fine, but the class naming would be crazy. Also numbers can't start namespace section (`Post/2016/` is not allowed).

So I suggest:

### 3. Use id option in post config

Class name => Directory location

- `Posts/Year2016/Post1` => `src/Posts/Year2016/Post1.php`

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+### Event
+
+**This is name of a place**. When something has happened in application: *order is sent*, 
+or *user is deleted*.     
+
+
+### Event Subscriber
+
+This **the action that happens** when we come to some place. When order is sent, *send me a confirmation sms*. And check storage they have all the ordered products we need. This means, that 1 event can invoke MANY different Event Subscribers.
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: Remove the "This" at the beginning or add "is".

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+final class NotifyMeOnVideoPublishedEventSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var bool
+     */
+    public $isUserNotified = false;
+
+    public static function getSubscribedEvents() : array
+    {
+        // in format ['event.name' => 'public function name that will be called']
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: `'event name' => 'public method to be called'`

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+    public function notifyUserAboutVideo()
+    {
+        // some logic to send notification
+        $this->isUserNotified = true;
+    }
+}
+```
+
+And add Subscriber to Dispatcher. Without that, he doesn't know about it.
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: "he", "it"... I don't like that, what about this? "Let the Dispatcher know about the Subscriber"

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+```language-php
+$eventDispatcher = new Symfony\Component\EventDispatcher\EventDispatcher;
+$notifyMeOnVideoPublishedEventSubscriber = new NotifyMeOnVideoPublishedEventSubscriber;
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: Too long variable imo, I'd use just `$subscriber` or `$eventSubscriber`.

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+---
+
+Still on? Let's get advanced.
+
+What if we need to get the name of the Youtuber into the Subscriber?
+
+
+## Event Objects to the Rescue!
+
+The Event objects are basically Value Objects. Pass a value in constructor a get with getter.
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: "and get it with getter" ?

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+
+### 1. Create an Event Object
+
+```language-php
+use Symfony\Component\EventDispatcher\Event;
+
+final class YoutuberNameEvent extends Event
```
Comment (enumag) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: This is not really related to the article itself but personally I prefer simple value objects to have public properties and no methods to keep them simple. I'd like to know your opinion about this topic.

**Comment by enumag** [comment]
@TomasVotruba I see you've chosen to combine 3 and 4. Personally I'd go with just 4 but it's not too important.

**Comment by EmanueleMinotto** [comment]
About the article 2016-12-05-symfony-event-dispatcher.md

1. the first section "2 main Features of Event Dispatcher" isn't so clear, I don't get those two points; the rest of the article is clear
2. imho suggesting also some best practices (e.g. the final class Events containing events string), could differentiate it from the others available on the net

my 2 cents :)

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
+---
+layout: post
+title: "Event Dispatcher from the Scratch"
+perex: "Today we look at first Symfony component - Event Dispatcher. Why start with it? It gives you flexibility, it is easy to understand and it helps you to write decoupled code."
```
Comment (petrvacha) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: Today we **are** look**ing** at

 Why **should you** start with it?

**a** decoupled code

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 2 above)
Comment (petrvacha) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: **The event subscriber describes an action which we want to trigger. E.g.** when order is sent,

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 5 above)
Comment (petrvacha) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: I would use sth shorter but with a descriptive name. What about `$videoPublishedEventSubscriber`?

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 7 above)
Comment (petrvacha) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: @enumag http://www.yegor256.com/2014/09/16/getters-and-setters-are-evil.html

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 2 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: I improved the wording here. Thanks

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 3 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: Good. More explicit, ok

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 5 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: I see. This might slow down reading and weaken the reference.
`$eventSubscriber` is just fine here, since they are added on background by DI Container later.

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 6 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: @enumag Exactly what I meant, thanks.

@EmanueleMinotto Good point! Do you have some resources to link to?

Found this, but seems to long: http://richardmiller.co.uk/2014/11/06/value-objects/ Same for others.

I will use that for now. We can replace it later with better one.

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md (hunk 7 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: @enumag It's difficult to pass the knowledge where to use public properties or where not. People tend to use what is shown everywhere (or many irrelevant places), which consequently breaks the SOLID principles (container in presenter/controller, inheritance over composition etc.). Thus I'd be strict and clear here, not to cause confusion.

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    // Event Object is passed as method argument
+    public function notifyUserAboutVideo(YoutuberNameEvent $youtuberNameEvent)
+    {
+        var_dump($youtuberNameEvent->getYoutubeName());    
+    }
+}
+```
+
+
+### 3. Create and Object and Dispatch With It
```
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: Thanks

File: source/_posts/2016/2016-12-05-symfony-event-dispatcher.md [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+```language-php
+$youtuberNameEvent = new YoutuberNameEvent('Jirka Král');
```
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-05-symfony-event-dispatcher.md]: I'd rather not cause any wars :D

**Comment by TomasVotruba** [comment]
@EmanueleMinotto I tried to improve it a bit and added 2 real use cases.
 It's 2 sides of one coin and it might be confusing, because it is actually one feature.

As for constant names in events: do you mean something like this?
http://symfony.com/doc/current/components/event_dispatcher.html#creating-an-event-class

I've added them to single Events class before (as seen in other packages, with no real thinking) and this looks better. Related to event class. What do you think?

**Comment by TomasVotruba** [comment]
@enumag Option 4 seems the best for me as well. I just needed to try it practice to see. Nothing is permanent.

Thank you for your feedback.

**Comment by TomasVotruba** [comment]
I'm merging this now, because there are lot of blog related changes that needs to be finished ASAP.

**Please, feel free to comment further or open PR for changes.**
